### PR TITLE
feat(inference): add batch geomodel inference for range filter

### PR DESCRIPTION
**Pull request by tphakala**
## Summary

- Add `BatchRangeFilter` interface and `PredictBatchRaw` method to the ONNX range filter, enabling the upcoming heatmap API to compute species scores across thousands of grid points in a single `session.Run()` call instead of one per point
- Follow the same pattern as `Classifier.PredictBatch`: flat input slice, single tensor allocation, copy output before tensor destruction
- Add `ErrSessionClosed` sentinel and nil-session guards to all inference methods across `RangeFilter`, `Classifier`, and `CustomClassifier` (previously panicked on nil pointer dereference after `Close()`)
- Fix pre-existing race in `DestroyONNXRuntime` where the mutex was unlocked before `ort.DestroyORT()` completed
- Refactor `filterBatchPredictions` to build the scoreMap once instead of rebuilding per batch iteration

Part of the Migration Explorer feature (#3092), PR 1 of 4.

## Test plan

- [x] Input validation tests: empty batch, negative batch size, length mismatch
- [x] Session-closed guard tests for all three RangeFilter methods
- [x] Error message format tests
- [x] Compile-time interface satisfaction check (`var _ BatchRangeFilter`)
- [x] All 580+ existing tests pass with `-race`
- [x] `golangci-lint run -v` clean

## Summary by CodeRabbit

* **New Features**
  * Added batch inference for range-based predictions to process multiple locations/weeks in one request.

* **Bug Fixes**
  * Improved handling when inference sessions are closed and when batch inputs are invalid, returning clear errors instead of runtime failures.

* **Tests**
  * Added unit tests for batch prediction validation, input/output length checks, and session-closed behavior.

[Review Change Stack](https://app.coderabbit.ai/change-stack/tphakala/birdnet-go/pull/3105?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: internal/inference/onnx.go [hunk 1]
```diff
@@ -213,6 +216,13 @@ func (r *onnxRangeFilter) Predict(latitude, longitude, week float32) ([]float32,
 	return r.filter.PredictRaw(latitude, longitude, week)
 }
 
+// PredictBatch runs batch inference on multiple location/week inputs.
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx.go:237]: Valid catch. The inner session-closed guards in `ort.RangeFilter` etc. are unreachable because the adapter dereferences `r.filter` first. Fixed in 2591f60 by adding nil guards at the adapter level (`onnxRangeFilter`, `onnxClassifier`, `onnxCustomClassifier`) that return `ErrSessionClosed` before dereferencing the inner object.

File: internal/inference/onnx.go [hunk 2]
```diff
@@ -213,6 +216,13 @@ func (r *onnxRangeFilter) Predict(latitude, longitude, week float32) ([]float32,
...
+// inputs is a flat slice of [lat, lon, week] triples: len(inputs) must equal batchSize * 3.
+// Returns a flat slice of [batchSize * numSpecies] scores in row-major order.
+func (r *onnxRangeFilter) PredictBatch(inputs []float32, batchSize int) ([]float32, error) {
+	return r.filter.PredictBatchRaw(inputs, batchSize)
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx.go:245]: Good catch. Fixed in 2591f60: added `r.filter == nil` guard in both `Predict` and `PredictBatch` at the adapter level, returning `ort.ErrSessionClosed` before dereferencing.

File: internal/inference/onnx/classifier.go [hunk 3]
```diff
@@ -258,6 +258,9 @@ func (c *Classifier) PredictRaw(audio []float32) ([]float32, error) {
 // PredictRawWithEmbeddings runs inference and returns both raw logits and embedding vector.
 // Returns nil embeddings if the model does not produce embeddings.
 func (c *Classifier) PredictRawWithEmbeddings(audio []float32) (logits, embeddings []float32, err error) {
+	if c.session == nil {
+		return nil, nil, ErrSessionClosed
+	}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx/classifier.go:263]: These structs are explicitly documented as NOT goroutine-safe (`backend.go` interface docs). Callers must synchronize access; the locking is handled at the BirdNET struct level, not at the ONNX struct level. The nil check is defense-in-depth for use-after-close, not a synchronization mechanism. Adding mutexes here would be redundant with the caller-side locking and add unnecessary overhead to every inference call.

File: internal/inference/onnx/custom_classifier.go [hunk 4]
```diff
@@ -146,6 +146,9 @@ func (c *CustomClassifier) Predict(embeddings []float32) ([]Prediction, error) {
 // PredictRaw runs inference on a single embedding vector and returns all
 // sigmoid-applied scores as a flat array (no topK filtering).
 func (c *CustomClassifier) PredictRaw(embeddings []float32) ([]float32, error) {
+	if c.session == nil {
+		return nil, ErrSessionClosed
+	}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx/custom_classifier.go:151]: Same as above: goroutine safety is the caller's responsibility per the documented interface contract. The nil check guards against use-after-close, not concurrent access.

File: internal/inference/onnx.go [hunk 5]
```diff
@@ -156,6 +162,9 @@ func NewONNXCustomClassifier(modelPath string, opts ONNXCustomClassifierOptions)
 
 // PredictEmbedding runs inference on an embedding vector.
 func (c *onnxCustomClassifier) PredictEmbedding(embeddings []float32) ([]float32, error) {
+	if c.classifier == nil {
+		return nil, ort.ErrSessionClosed
+	}
 	return c.classifier.PredictRaw(embeddings)
 }
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/inference/onnx.go:170]: Valid. Fixed in latest commit: added nil guards to NumClasses() and Labels() on onnxCustomClassifier, returning zero values when the classifier is closed.
